sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PullOutJoinConditionSuite.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.{Alias, Coalesce, IsNull, Literal, Substring, Upper}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules._
+import org.apache.spark.sql.internal.SQLConf
+
+class PullOutJoinConditionSuite extends PlanTest {
+
+  private object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches =
+      Batch("Pull out join condition", Once,
+        PullOutJoinCondition,
+        CollapseProject) :: Nil
+  }
+
+  private val testRelation = LocalRelation('a.string, 'b.int, 'c.int)
+  private val testRelation1 = LocalRelation('d.string, 'e.int)
+  private val x = testRelation.subquery('x)
+  private val y = testRelation1.subquery('y)
+
+  test("Pull out join keys evaluation(String expressions)") {
+    val joinType = Inner
+    Seq(Upper("y.d".attr), Substring("y.d".attr, 1, 5)).foreach { udf =>
+      val originalQuery = x.join(y, joinType, Option("x.a".attr === udf))
+        .select("x.a".attr, "y.e".attr)
```

Comment:
we don't need to use qualified column names. There is no name conflict.

Reply:
ditto for other places in this test suite

Reply:
Removed it. But UDF need it otherwise:
```
== FAIL: Plans do not match ===
 'Project [a#0, e#0]                                    'Project [a#0, e#0]
!+- 'Join Inner, (a#0 = upper(y.d)#0)                   +- 'Join Inner, (upper(d)#0 = a#0)
    :- Project [a#0, b#0, c#0]                             :- Project [a#0, b#0, c#0]
    :  +- LocalRelation <empty>, [a#0, b#0, c#0]           :  +- LocalRelation <empty>, [a#0, b#0, c#0]
!   +- Project [d#0, e#0, upper(d#0) AS upper(y.d)#0]      +- Project [d#0, e#0, upper(d#0) AS upper(d)#0]
       +- LocalRelation <empty>, [d#0, e#0]                   +- LocalRelation <empty>, [d#0, e#0]
  
```